File: app/models/wellness_checkin.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import (
    Column, String, DateTime, Date, SmallInteger, Text, ForeignKey, Index,
)
from sqlalchemy.dialects.postgresql import UUID, ARRAY

from app.database import Base


# ── Entry methods ──────────────────────────────────────────────────────────
ENTRY_MORNING = "morning"
ENTRY_PRE_RUN = "pre_run"
ENTRY_POST_RUN = "post_run"
ENTRY_MANUAL = "manual"


class WellnessCheckin(Base):
    """
    Subjective wellness data — sliders + optional body-part chips + free-text.

    Phase 4 introduces three flavors via `entry_method`:
      - morning: full check-in, 4 sliders + body areas + notes. ONE per day per user.
      - pre_run: 3 quick sliders (sleep / energy / body) on Start Run.
        Multiple per day allowed.
      - post_run / manual: future hooks; same shape, no day-uniqueness constraint.

    Activates `pain_logged` anomaly flag in Phase 3's anomaly_engine and feeds
    weekly review prompt input.
    """
    __tablename__ = "wellness_checkins"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    date = Column(Date, nullable=False)
    entry_method = Column(String(16), nullable=False)  # ENTRY_* values

    # Sliders 1-5. Nullable so partial entries are allowed (pre_run skips stress/motivation).
    sleep_quality = Column(SmallInteger, nullable=True)
    soreness = Column(SmallInteger, nullable=True)
    motivation = Column(SmallInteger, nullable=True)
    stress = Column(SmallInteger, nullable=True)
    energy = Column(SmallInteger, nullable=True)  # set on pre_run only

    # Optional structured body-area chips (e.g., ["calves", "left_knee"]).
    soreness_areas = Column(ARRAY(String), nullable=False, default=list)

    # Optional free-text. iOS dictation-enabled.
    notes = Column(Text, nullable=True)

    submitted_at = Column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )

    __table_args__ = (
        # The unique-per-day constraint on morning entries lives on the partial
        # index added in main.py inline migrations (SQLAlchemy doesn't model
        # partial-unique-indices cleanly across dialects).
        Index("ix_wellness_user_date", "user_id", "date"),
        Index("ix_wellness_user_method_date", "user_id", "entry_method", "date"),
    )

File: StrideApp/Models/WellnessCheckin.swift
import Foundation
import SwiftData

/// Local SwiftData mirror of backend wellness_checkins for offline display +
/// quick re-render on Stats / Run tabs without an extra fetch.
///
/// The server is the source of truth. iOS submits via POST /api/wellness/checkin
/// and inserts the returned row here. Inbox + trends come from /today, /trends,
/// /history endpoints.
@Model
final class WellnessCheckin {
    @Attribute(.unique) var id: String

    /// Server-stamped date of the entry (yyyy-MM-dd).
    var date: String

    /// "morning" | "pre_run" | "post_run" | "manual"
    var entryMethod: String

    var sleepQuality: Int?
    var soreness: Int?
    var motivation: Int?
    var stress: Int?
    var energy: Int?

    var sorenessAreas: [String]
    var notes: String?

    var submittedAt: Date

    init(
        id: String,
        date: String,
        entryMethod: String,
        sleepQuality: Int? = nil,
        soreness: Int? = nil,
        motivation: Int? = nil,
        stress: Int? = nil,
        energy: Int? = nil,
        sorenessAreas: [String] = [],
        notes: String? = nil,
        submittedAt: Date
    ) {
        self.id = id
        self.date = date
        self.entryMethod = entryMethod
        self.sleepQuality = sleepQuality
        self.soreness = soreness
        self.motivation = motivation
        self.stress = stress
        self.energy = energy
        self.sorenessAreas = sorenessAreas
        self.notes = notes
        self.submittedAt = submittedAt
    }
}
